injection/sharedmain/main.go

```diff
@@ -143,101 +166,179 @@ func MainWithConfig(ctx context.Context, component string, cfg *rest.Config, cto
 	defer flush(logger)
 	ctx = logging.WithLogger(ctx, logger)
 
-	// Obtain K8s clientset.
-	kc := kubeclient.Get(ctx)
-	if err := version.CheckMinimumVersion(kc.Discovery()); err != nil {
-		logger.Fatalw("Version check failed", zap.Error(err))
-	}
+	// run is the main controller flow
+	run := func(ctx context.Context) {
```

Comment:
This is quite big. Why not extract it to a separate function?

Reply:
I was following the constructions in k8s controller manager, decided to save any cosmetics or discretionary factoring for after it worked. Since that function captures many variables from the enclosing context, extracting a function would be more laborious than it might appear - since you would have to introduce a type or function factory to inject those variables.

Reply:
I didn't look closely, but I noticed logger. If it's just 2-3, then passing via args should be alright.
But, sure, we can brush it up later on.

Reply:
Frankly, I don't think it's worth it, but you're welcome to do it as a follow-on PR.